src/driver-steelseries.c

```diff
@@ -421,14 +423,15 @@ steelseries_probe(struct ratbag_device *device)
 		}
 	}
 
-	drv_data = zalloc(sizeof(*drv_data));
-	ratbag_set_drv_data(device, drv_data);
-
-	rc = steelseries_get_firmware_version(device);
-	if(rc == 0)
-		log_debug(device->ratbag, "SteelSeries firmware version %d.%d\n",
-			drv_data->firmware_major,
-			drv_data->firmware_minor);
+	{
+		int firmware_minor = 0;
+		int firmware_major = 0;
+		rc = steelseries_get_firmware_version(device, &firmware_minor, &firmware_major);
+		if (rc == 0) {
+			device->firmware_version = zalloc(STEELSERIES_FIRMWARE_VERSION_STRING_LENGTH + 1);
+			snprintf(device->firmware_version, STEELSERIES_FIRMWARE_VERSION_STRING_LENGTH + 1, "%d.%d", firmware_major, firmware_minor);
```

Comment:
this could be a simpler:
```c
device->firmware_version = asprintf_safe("%d.%d", major, minor);
```

(`asprintf_safe` should probably  be changed to return `char*` but that's separate)

Reply:
In such cases would it be preferable to set `device->firmware_version` directly to not do a redundant string copy, or should the `ratbag_device_set_fw_version` API be used to stay on higher abstraction layer?

Reply:
the cost of doing this is miniscule - you're doing this exactly once during device startup and a string copy isn't expensive in the grand scheme of things. So something like this is best, IMO:
```c
_cleanup_free_ char *fw = asprintf_safe("%d.%d", major, minor);
ratbag_device_set_fw_version(device, fw);
```
(or whatever the right cleanup thing is)

alternatively and I only realised this now: how about making `ratbag_device_set_fw_version()` a printf-like function? Then most drivers can just do `ratbag_device_set_fw_version(device, "%d.%d", major, minor)` and be done with it.

Reply:
> So something like this is best, IMO

That's what I will do, then. :)

Still, I'm not a fan of doing any wasteful operations, but here it really isn't heavy at all and having the code more maintainable is more important.

> how about making `ratbag_device_set_fw_version()` a printf-like function

I don't really like this, as this would make the function do several things at the same time. Having a separate formatting and data setting is good as is in my opinion.